### PR TITLE
DeleteWithOutput, DeleteWithOutputInto for MSSQL.

### PR DESCRIPTION
**Pull request by viceroypenguin**

### CODE REVIEW

File: Source/LinqToDB/LinqExtensions.Delete.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+			return currentSource.Provider.CreateQuery<TSource>(
+					Expression.Call(
+						null,
+						MethodHelper.GetMethodInfo(DeleteWithOutput, source),
+						currentSource.Expression))
+				.AsEnumerable();
+		}
+
+		/// <summary>
+		/// Deletes records from source query into target table asynchronously and returns newly created records.
```
Comment (sdanyliv) [inline Source/LinqToDB/LinqExtensions.Delete.cs:39]: I think
```suggestion
		/// Deletes records from source query into target table asynchronously and returns returns deleted records.
```

File: Source/LinqToDB/LinqExtensions.Delete.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+			return currentSource.Provider.CreateQuery<TSource>(
+					Expression.Call(
+						null,
+						MethodHelper.GetMethodInfo(DeleteWithOutput, source),
+						currentSource.Expression))
+				.ToArrayAsync(token);
+		}
+
+		/// <summary>
+		/// Deletes records from source query into target table and returns newly created records.
```
Comment (sdanyliv) [inline Source/LinqToDB/LinqExtensions.Delete.cs:62]:
```suggestion
		/// Deletes records from source query into target table and returns deleted records.
```

File: Source/LinqToDB/LinqExtensions.Delete.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+					Expression.Call(
+						null,
+						MethodHelper.GetMethodInfo(DeleteWithOutput, source, outputExpression),
+						currentSource.Expression,
+						Expression.Quote(outputExpression)))
+				.AsEnumerable();
+		}
+
+		/// <summary>
+		/// Deletes records from source query into target table asynchronously and returns newly created records.
```
Comment (sdanyliv) [inline Source/LinqToDB/LinqExtensions.Delete.cs:90]:
```suggestion
		/// Deletes records from source query into target table asynchronously and returns deleted records.
```

File: Source/LinqToDB/LinqExtensions.Delete.cs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+					Expression.Call(
+						null,
+						MethodHelper.GetMethodInfo(DeleteWithOutput, source, outputExpression),
+						currentSource.Expression,
+						Expression.Quote(outputExpression)))
+				.ToArrayAsync(token);
+		}
+
+		/// <summary>
+		/// Deletes records from source query into target table and outputs newly created records into <paramref name="outputTable"/>.
```
Comment (sdanyliv) [inline Source/LinqToDB/LinqExtensions.Delete.cs:119]:
```suggestion
		/// Deletes records from source query into target table and outputs deleted records into <paramref name="outputTable"/>.
```

File: Tests/Linq/Update/DeleteWithOutputTests.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		public void DeleteWithOutputProjectionFromQueryTest([IncludeDataSources(true, TestProvName.AllSqlServer2008Plus)] string context, [Values(100, 200)] int param)
+		{
+			var sourceData    = GetSourceData();
+			using (var db     = GetDataContext(context))
+			using (var source = db.CreateLocalTable(sourceData))
+			{
+				var zz = source
+					.Where(s => s.Id > 3)
+					.ToArray();
+
```
Comment (sdanyliv) [inline Tests/Linq/Update/DeleteWithOutputTests.cs:57]: I think we have to remove that
```suggestion
```

File: Tests/Linq/Update/DeleteWithOutputTests.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		public void DeleteWithOutputFromQueryTest([IncludeDataSources(true, TestProvName.AllSqlServer2008Plus)] string context, [Values(100, 200)] int param)
+		{
+			var sourceData    = GetSourceData();
+			using (var db     = GetDataContext(context))
+			using (var source = db.CreateLocalTable(sourceData))
+			{
+				var zz = source
+					.Where(s => s.Id > 3)
+					.ToArray();
+
```
Comment (sdanyliv) [inline Tests/Linq/Update/DeleteWithOutputTests.cs:87]: I think we have to remove that
```suggestion
```

File: Tests/Linq/Update/DeleteWithOutputTests.cs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		public async Task DeleteWithOutputFromQueryTestAsync([IncludeDataSources(true, TestProvName.AllSqlServer2008Plus)] string context, [Values(100, 200)] int param)
+		{
+			var sourceData    = GetSourceData();
+			using (var db     = GetDataContext(context))
+			using (var source = db.CreateLocalTable(sourceData))
+			{
+				var zz = source
+					.Where(s => s.Id > 3)
+					.ToArray();
+
```
Comment (sdanyliv) [inline Tests/Linq/Update/DeleteWithOutputTests.cs:119]: I think we have to remove that
```suggestion
```

File: Tests/Linq/Update/DeleteWithOutputTests.cs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		{
+			var sourceData    = GetSourceData();
+			using (var db     = GetDataContext(context))
+			using (var source = db.CreateLocalTable(sourceData))
+			using (var target = db.CreateLocalTable<DestinationTable>())
+			{
+				var zz = source
+					.Where(s => s.Id > 3)
+					.ToArray();
+
```
Comment (sdanyliv) [inline Tests/Linq/Update/DeleteWithOutputTests.cs:151]: I think we have to remove that
```suggestion
```

File: Tests/Linq/Update/DeleteWithOutputTests.cs (hunk 5 above)
Comment (viceroypenguin) [reply Tests/Linq/Update/DeleteWithOutputTests.cs:57]: `zz` is used in the `AreEqual()` at the end of the function. This needs to be done first so we have the values in memory for comparison before deleting them. Same for the others.

**Comment by sdanyliv** [comment]
Ah, ok. I use `zz` in my code to create temporary variables which has to be removed (if i forgot about them). Maybe just give better name.

**Comment by viceroypenguin** [comment]
Yup. Copied what I saw in InsertWithOutputTests without thinking about it. All good. 😊

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by sdanyliv** [comment]
Anyway, nice extraction of needed functionality ;)

**Pull request merged**